hack/jenkins/e2e-runner.sh

```diff
@@ -21,8 +21,53 @@ set -o nounset
 set -o pipefail
 set -o xtrace
 
+# include shell2junit library
+source <(curl -fsS --retry 3 'https://raw.githubusercontent.com/kubernetes/kubernetes/master/third_party/forked/shell2junit/sh2ju.sh')
+
+# Have cmd/e2e run by goe2e.sh generate JUnit report in ${WORKSPACE}/junit*.xml
+ARTIFACTS=${WORKSPACE}/_artifacts
+mkdir -p ${ARTIFACTS}
+
+# E2E runner stages
+STAGE_PRE="PRE-SETUP"
+STAGE_SETUP="SETUP"
+STAGE_CLEANUP="CLEANUP"
+STAGE_KUBEMARK="KUBEMARK"
+
 : ${KUBE_GCS_RELEASE_BUCKET:="kubernetes-release"}
 
+# record_command records the command output and error message in junit format
+# it expects the first argument to be the class and the second to be the name of the command
+# Example:
+# record_command PRESETUP curltest curl google.com
+# record_command CLEANUP check false
+#
+# WARNING: variable changes will NOT be effective after record_command returns
+function record_command() {
+    set +o xtrace
+    set +o nounset
+    set +o errexit
+
+    class=$1
+    shift
+    name=$1
+    shift
+    echo "Recording: ${class} ${name}"
+    echo "Running command: $@"
+    local -r ret=$(juLog -output="${ARTIFACTS}" -class="${class}" -name="${name}" "$@")
+
+    set -o nounset
+    set -o errexit
+    set -o xtrace
+
+    # since record_command disable errexit, exit script on failure.
+    local -r found=$(echo "${ret}" | grep -c "+++ exit code: 0")
+    if [ ${found} -le 0 ]; then
```

Comment:
This PR LGTM when bash is good :)

On Tue, Jun 28, 2016 at 1:12 PM, Minhan Xia <notifications@github.com>
wrote:

> In hack/jenkins/e2e-runner.sh
> <https://github.com/kubernetes/kubernetes/pull/28133#discussion_r68833550>
> :
>
> > +
> > +    class=$1
> > +    shift
> > +    name=$1
> > +    shift
> > +    echo "Recording: ${class} ${name}"
> > +    echo "Running command: $@"
> > +    local -r ret=$(juLog -output="${ARTIFACTS}" -class="${class}" -name="${name}" "$@")
> > +
> > +    set -o nounset
> > +    set -o errexit
> > +    set -o xtrace
> > +
> > +    # since record_command disable errexit, exit script on failure.
> > +    local -r found=$(echo "${ret}" | grep -c "+++ exit code: 0")
> > +    if [ ${found} -le 0 ]; then
>
> juLog returns a bunch of stuff. I have to ensure the output indicates the
> command is successful.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28133/files/0b2922834a2adc130539f915f2ffb23f8e404f71#r68833550>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAnglilVIzESJzhmp8b2rcV-pjRlsRPXks5qQYAqgaJpZM4I_jtj>
> .
>